src/mlpack/methods/kde/kde_main.cpp

```diff
@@ -0,0 +1,189 @@
+  const std::string treeStr = CLI::GetParam<std::string>("tree");
+  const double relError = CLI::GetParam<double>("rel_error");
+  const double absError = CLI::GetParam<double>("abs_error");
+  // Initialize results vector.
+  arma::vec estimations;
+
+  // You can only specify reference data or a pre-trained model.
+  RequireOnlyOnePassed({ "reference", "input_model" }, true);
+  ReportIgnoredParam({{ "input_model", true }}, "tree");
+  ReportIgnoredParam({{ "input_model", true }}, "kernel");
+  ReportIgnoredParam({{ "input_model", true }}, "rel_error");
+  ReportIgnoredParam({{ "input_model", true }}, "abs_error");
+
+  // Requirements for parameter values.
+  RequireParamInSet<string>("kernel", { "gaussian", "epanechnikov",
+      "laplacian", "spherical", "triangular" }, true, "unknown kernel type");
+  RequireParamInSet<string>("tree", { "kd-tree", "ball-tree", "cover-tree",
+      "octree", "r-tree"}, true, "unknown tree type");
+  RequireParamValue<double>("rel_error", [](double x){return x >= 0 && x <= 1;},
+      true, "relative error must be between 0 and 1");
+  RequireParamValue<double>("abs_error", [](double x){return x >= 0;},
+      true, "absolute error must be equal or greater than 0");
+
+  KDEModel* kde;
+
+  if (CLI::HasParam("reference"))
+  {
+    arma::mat reference = std::move(CLI::GetParam<arma::mat>("reference"));
+
+    kde = new KDEModel();
+    // Set parameters
+    kde->Bandwidth() = bandwidth;
+    kde->RelativeError() = relError;
+    kde->AbsoluteError() = absError;
+
+    // Set KernelType
+    if (kernelStr == "gaussian")
+      kde->KernelType() = KDEModel::GAUSSIAN_KERNEL;
+    else if (kernelStr == "epanechnikov")
+      kde->KernelType() = KDEModel::EPANECHNIKOV_KERNEL;
+    else if (kernelStr == "laplacian")
+      kde->KernelType() = KDEModel::LAPLACIAN_KERNEL;
+    else if (kernelStr == "spherical")
+      kde->KernelType() = KDEModel::SPHERICAL_KERNEL;
+    else if (kernelStr == "triangular")
+      kde->KernelType() = KDEModel::TRIANGULAR_KERNEL;
+
+    // Set TreeType
+    if (treeStr == "kd-tree")
+      kde->TreeType() = KDEModel::KD_TREE;
+    else if (treeStr == "ball-tree")
+      kde->TreeType() = KDEModel::BALL_TREE;
+    else if (treeStr == "cover-tree")
+      kde->TreeType() = KDEModel::COVER_TREE;
+    else if (treeStr == "octree")
+      kde->TreeType() = KDEModel::OCTREE;
+    else if (treeStr == "r-tree")
+      kde->TreeType() = KDEModel::R_TREE;
+
+    // Build model
+    kde->BuildModel(std::move(reference));
+  }
+  else
+  {
+    // Load model
+    kde = CLI::GetParam<KDEModel*>("input_model");
+  }
+
+  if (CLI::HasParam("query"))
+  {
+    arma::mat query = std::move(CLI::GetParam<arma::mat>("query"));
+    kde->Evaluate(std::move(query), estimations);
+  }
+  else
+    kde->Evaluate(estimations);
+
+  // Output results if needed.
+  if (CLI::HasParam("output"))
+    CLI::GetParam<arma::mat>("output") = std::move(estimations);
```

Comment:
We should transpose this---when I run this I get a big CSV with one line and all the predictions on that line.  Also maybe it would be better to call the output `predictions` or `estimates`?

Reply:
Solved in 0bb023aba67d0550db02450e9dd59ec700072d5c